File: cpp/src/scene.cpp
#include "scene.h"
#include <algorithm>
#include <cmath>
#include <stack>
#include <sstream>

namespace ocp {

// SceneNode
SceneNode* SceneNode::add_child(std::unique_ptr<SceneNode> child) {
    child->parent = this;
    SceneNode* raw = child.get();
    children.push_back(std::move(child));
    return raw;
}

void SceneNode::remove_child(SceneNode* child) {
    if (!child) return;
    child->parent = nullptr;
    children.erase(
        std::remove_if(children.begin(), children.end(),
            [child](const std::unique_ptr<SceneNode>& c) { return c.get() == child; }),
        children.end());
}

void SceneNode::update_matrices(const mat4& parent_world) {
    mat4 T = glm::translate(mat4(1.0f), position);
    mat4 Rx = glm::rotate(mat4(1.0f), deg2rad(rotation.x), vec3(1, 0, 0));
    mat4 Ry = glm::rotate(mat4(1.0f), deg2rad(rotation.y), vec3(0, 1, 0));
    mat4 Rz = glm::rotate(mat4(1.0f), deg2rad(rotation.z), vec3(0, 0, 1));
    mat4 S = glm::scale(mat4(1.0f), scale);
    local_matrix = T * Ry * Rx * Rz * S;
    world_matrix = parent_world * local_matrix;
    for (auto& c : children) c->update_matrices(world_matrix);
}

vec3 SceneNode::get_bounding_box_min() const {
    if (mesh && !mesh->vertices.empty()) {
        vec3 mn = mesh->get_bounding_box_min();
        mn = vec3(world_matrix * vec4(mn, 1.0f));
        return mn;
    }
    vec3 mn = position;
    for (auto& c : children) {
        vec3 cmn = c->get_bounding_box_min();
        mn = glm::min(mn, cmn);
    }
    return children.empty() ? position - glm::abs(scale) : mn;
}

vec3 SceneNode::get_bounding_box_max() const {
    if (mesh && !mesh->vertices.empty()) {
        vec3 mx = mesh->get_bounding_box_max();
        mx = vec3(world_matrix * vec4(mx, 1.0f));
        return mx;
    }
    vec3 mx = position;
    for (auto& c : children) {
        vec3 cmx = c->get_bounding_box_max();
        mx = glm::max(mx, cmx);
    }
    return children.empty() ? position + glm::abs(scale) : mx;
}

std::unique_ptr<SceneNode> SceneNode::clone_recursive() const {
    auto n = std::make_unique<SceneNode>();
    n->name = name + "_copy";
    n->position = position;
    n->rotation = rotation;
    n->scale = scale;
    n->visible = visible;
    n->locked = locked;
    if (mesh) n->mesh = std::make_unique<Mesh>(mesh->clone());
    if (material) {
        n->material = std::make_unique<Material>();
        *n->material = *material;
    }
    for (auto& c : children) {
        SceneNode* cc = n->add_child(c->clone_recursive());
        (void)cc;
    }
    return n;
}

void SceneNode::delete_children() {
    for (auto& c : children) {
        c->delete_children();
    }
    children.clear();
}

SceneNode* SceneNode::find_child(const std::string& nm) const {
    for (auto& c : children) {
        if (c->name == nm) return c.get();
        SceneNode* found = c->find_child(nm);
        if (found) return found;
    }
    return nullptr;
}

// Scene
Scene::Scene() {
    root.name = "Root";
}

SceneNode* Scene::add_node(std::unique_ptr<SceneNode> node, SceneNode* parent) {
    if (!parent) parent = &root;
    SceneNode* raw = parent->add_child(std::move(node));
    return raw;
}

SceneNode* Scene::add_node_raw(SceneNode* node, SceneNode* parent) {
    if (!parent) parent = &root;
    std::unique_ptr<SceneNode> ptr(node);
    return parent->add_child(std::move(ptr));
}

void Scene::remove_node(SceneNode* node) {
    if (!node || node == &root) return;
    if (node->parent) node->parent->remove_child(node);
    if (selected_node == node) selected_node = nullptr;
    multi_selection.erase(
        std::remove(multi_selection.begin(), multi_selection.end(), node),
        multi_selection.end());
}

std::vector<SceneNode*> Scene::get_all_nodes() const {
    std::vector<SceneNode*> result;
    std::stack<SceneNode*> stk;
    for (auto& c : root.children) stk.push(c.get());
    while (!stk.empty()) {
        SceneNode* n = stk.top(); stk.pop();
        result.push_back(n);
        for (auto& c : n->children) stk.push(c.get());
    }
    return result;
}

SceneNode* Scene::find_node(const std::string& nm) const {
    for (auto* n : get_all_nodes()) {
        if (n->name == nm) return n;
    }
    return nullptr;
}

int Scene::get_node_count() const {
    return (int)get_all_nodes().size();
}

void Scene::update() {
    root.update_matrices();
}

void Scene::select(SceneNode* node) {
    selected_node = node;
    multi_selection.clear();
    if (node) multi_selection.push_back(node);
}

void Scene::toggle_multi_select(SceneNode* node) {
    if (!node) return;
    for (auto it = multi_selection.begin(); it != multi_selection.end(); ++it) {
        if (*it == node) {
            multi_selection.erase(it);
            if (selected_node == node) {
                selected_node = multi_selection.empty() ? nullptr : multi_selection.back();
            }
            return;
        }
    }
    multi_selection.push_back(node);
    selected_node = node;
}

void Scene::deselect() {
    selected_node = nullptr;
    multi_selection.clear();
}

void Scene::delete_selected() {
    if (selected_node) {
        remove_node(selected_node);
        selected_node = nullptr;
    }
}

void Scene::duplicate_selected() {
    if (!selected_node) return;
    auto dup = selected_node->clone_recursive();
    SceneNode* dup_raw = dup.get();
    dup_raw->position.x += 1.0f;
    if (selected_node->parent) {
        selected_node->parent->add_child(std::move(dup));
    } else {
        root.add_child(std::move(dup));
    }
    select(dup_raw);
}

void Scene::copy_selected() {
    clipboard.clear();
    for (auto* node : multi_selection) {
        clipboard.push_back(node->clone_recursive());
    }
    if (clipboard.empty() && selected_node) {
        clipboard.push_back(selected_node->clone_recursive());
    }
}

void Scene::paste_clipboard() {
    if (clipboard.empty()) return;
    std::vector<SceneNode*> new_nodes;
    for (auto& c : clipboard) {
        auto dup = c->clone_recursive();
        SceneNode* raw = dup.get();
        raw->position.x += 1.5f;
        root.add_child(std::move(dup));
        new_nodes.push_back(raw);
    }
    if (!new_nodes.empty()) {
        select(new_nodes.front());
    }
}

void Scene::group_selected() {
    if (multi_selection.size() < 2) return;
    auto group = std::make_unique<SceneNode>();
    group->name = "Group";
    SceneNode* group_raw = group.get();
    SceneNode* first_parent = multi_selection[0]->parent ? multi_selection[0]->parent : &root;
    first_parent->add_child(std::move(group));
    std::vector<SceneNode*> to_move = multi_selection;
    for (auto* node : to_move) {
        if (!node->parent) continue;
        vec3 world_pos = node->get_world_position();
        node->parent->remove_child(node);
        node->parent = nullptr;
        node->position = world_pos;
        auto ptr = std::unique_ptr<SceneNode>(node);
        group_raw->add_child(std::move(ptr));
    }
    select(group_raw);
}

void Scene::ungroup_selected() {
    if (!selected_node || selected_node->children.empty()) return;
    SceneNode* parent = selected_node->parent ? selected_node->parent : &root;
    vec3 group_world = selected_node->get_world_position();
    std::vector<std::unique_ptr<SceneNode>> kids;
    for (auto& c : selected_node->children) {
        vec3 child_world = c->get_world_position();
        c->position = child_world - group_world;
        kids.push_back(std::move(c));
    }
    selected_node->children.clear();
    SceneNode* to_remove = selected_node;
    for (auto& k : kids) {
        SceneNode* raw = k.get();
        parent->add_child(std::move(k));
    }
    parent->remove_child(to_remove);
    select(parent->children.empty() ? nullptr : parent->children.back().get());
}

void Scene::clear() {
    root.delete_children();
    cameras.clear();
    lights.clear();
    selected_node = nullptr;
    multi_selection.clear();
}

bool Scene::raycast(const vec3& origin, const vec3& direction,
                    SceneNode*& hit_node, float& hit_dist) const {
    hit_node = nullptr;
    hit_dist = 1e30f;
    for (auto& c : root.children) {
        float d;
        if (raycast_node(origin, direction, c.get(), d) && d < hit_dist) {
            hit_dist = d;
            hit_node = c.get();
        }
    }
    return hit_node != nullptr;
}

bool Scene::raycast_node(const vec3& origin, const vec3& direction,
                         SceneNode* node, float& dist) const {
    if (!node || !node->visible) return false;

    bool hit_this = false;
    float this_dist = 1e30f;

    if (node->mesh && node->mesh->get_vertex_count() > 0) {
        vec3 bmin = node->get_bounding_box_min();
        vec3 bmax = node->get_bounding_box_max();
        vec3 inv_dir = 1.0f / (direction + vec3(1e-8f));
        vec3 t1 = (bmin - origin) * inv_dir;
        vec3 t2 = (bmax - origin) * inv_dir;
        vec3 tmin = glm::min(t1, t2);
        vec3 tmax = glm::max(t1, t2);
        float t_near = std::max({tmin.x, tmin.y, tmin.z});
        float t_far = std::min({tmax.x, tmax.y, tmax.z});
        if (t_near <= t_far && t_far > 0) {
            this_dist = t_near > 0 ? t_near : t_far;
            hit_this = true;
        }
    }

    for (auto& c : node->children) {
        float child_dist;
        if (raycast_node(origin, direction, c.get(), child_dist)) {
            if (child_dist < this_dist) {
                this_dist = child_dist;
                hit_this = true;
                dist = this_dist;
                return true;
            }
        }
    }

    if (hit_this) {
        dist = this_dist;
        return true;
    }
    return false;
}

std::string Scene::save_to_string() const {
    std::ostringstream ss;
    ss << "OCP_SCENE_V1\n";
    ss << "name:" << name << "\n";
    ss << "bg:" << background_color.r << "," << background_color.g << ","
       << background_color.b << "," << background_color.a << "\n";
    ss << "grid:" << (grid_enabled ? 1 : 0) << "," << grid_size << "," << grid_subdivisions << "\n";

    ss << "cameras:" << cameras.size() << "\n";
    for (auto& c : cameras) {
        ss << "  " << c.name << "|" << c.position.x << "," << c.position.y << "," << c.position.z
           << "|" << c.target.x << "," << c.target.y << "," << c.target.z
           << "|" << c.fov << "\n";
    }

    ss << "lights:" << lights.size() << "\n";
    for (auto& l : lights) {
        ss << "  " << l.name << "|t:" << (int)l.type
           << "|p:" << l.position.x << "," << l.position.y << "," << l.position.z
           << "|d:" << l.direction.x << "," << l.direction.y << "," << l.direction.z
           << "|c:" << l.color.r << "," << l.color.g << "," << l.color.b
           << "|i:" << l.intensity << "|e:" << (l.enabled ? 1 : 0) << "\n";
    }

    std::function<void(const SceneNode&, int)> save_node = [&](const SceneNode& n, int depth) {
        ss << std::string(depth * 2, ' ') << "node:" << n.name
           << "|p:" << n.position.x << "," << n.position.y << "," << n.position.z
           << "|r:" << n.rotation.x << "," << n.rotation.y << "," << n.rotation.z
           << "|s:" << n.scale.x << "," << n.scale.y << "," << n.scale.z
           << "|v:" << (n.visible ? 1 : 0) << "\n";
        if (n.material) {
            ss << std::string(depth * 2 + 2, ' ') << "mat:"
               << n.material->to_hex()
               << "|m:" << n.material->metallic
               << "|r:" << n.material->roughness
               << "|sh:" << n.material->shininess
               << "|w:" << (n.material->wireframe ? 1 : 0)
               << "|o:" << n.material->opacity << "\n";
        }
        for (auto& c : n.children) save_node(*c, depth + 1);
    };

    ss << "nodes:" << root.children.size() << "\n";
    for (auto& c : root.children) save_node(*c, 0);

    return ss.str();
}

bool Scene::load_from_string(const std::string& data) {
    clear();
    std::istringstream ss(data);
    std::string line;
    std::getline(ss, line);
    if (line.find("OCP_SCENE_V1") == std::string::npos) return false;

    while (std::getline(ss, line)) {
        if (line.empty()) continue;
        if (line.substr(0, 5) == "name:") name = line.substr(5);
        else if (line.substr(0, 3) == "bg:") {
            sscanf(line.c_str(), "bg:%f,%f,%f,%f",
                &background_color.r, &background_color.g, &background_color.b, &background_color.a);
        }
        else if (line.substr(0, 5) == "grid:") {
            int g; sscanf(line.c_str(), "grid:%d,%f,%d", &g, &grid_size, &grid_subdivisions);
            grid_enabled = g != 0;
        }
        else if (line.substr(0, 8) == "lights:") {
            int count; sscanf(line.c_str(), "lights:%d", &count);
            for (int i = 0; i < count; i++) {
                std::getline(ss, line);
                Light l;
                int t, e;
                char name_buf[64] = {};
                sscanf(line.c_str(), "  %[^|]|t:%d|p:%f,%f,%f|d:%f,%f,%f|c:%f,%f,%f|i:%f|e:%d",
                    name_buf, &t, &l.position.x, &l.position.y, &l.position.z,
                    &l.direction.x, &l.direction.y, &l.direction.z,
                    &l.color.r, &l.color.g, &l.color.b, &l.intensity, &e);
                l.name = name_buf;
                l.type = (Light::Type)t;
                l.enabled = e != 0;
                lights.push_back(l);
            }
        }
    }
    return true;
}

} // namespace ocp

File: src/core/scene.py
import numpy as np
from ..utils.math_utils import mat4_identity, mat4_translate, mat4_scale, mat4_rotate_euler


class SceneNode:
    def __init__(self, name="Node", parent=None):
        self.name = name
        self.parent = parent
        self.children = []
        self.position = np.zeros(3, dtype=np.float32)
        self.rotation = np.zeros(3, dtype=np.float32)
        self.scale = np.ones(3, dtype=np.float32)
        self.visible = True
        self.locked = False
        self.mesh = None
        self.material = None
        self._local_matrix = mat4_identity()
        self._world_matrix = mat4_identity()
        self._dirty = True

    def add_child(self, child):
        child.parent = self
        self.children.append(child)

    def remove_child(self, child):
        child.parent = None
        self.children.remove(child)

    def get_children(self):
        return self.children

    def set_dirty(self):
        self._dirty = True
        for child in self.children:
            child.set_dirty()

    def update_matrices(self, parent_world=None):
        self._local_matrix = (
            mat4_translate(*self.position)
            @ mat4_rotate_euler(*np.radians(self.rotation))
            @ mat4_scale(*self.scale)
        )
        if parent_world is not None:
            self._world_matrix = parent_world @ self._local_matrix
        else:
            self._world_matrix = self._local_matrix.copy()
        self._dirty = False
        for child in self.children:
            child.update_matrices(self._world_matrix)

    def get_world_matrix(self):
        return self._world_matrix

    def get_local_matrix(self):
        return self._local_matrix

    def translate(self, dx, dy, dz):
        self.position += np.array([dx, dy, dz], dtype=np.float32)
        self.set_dirty()

    def rotate(self, rx, ry, rz):
        self.rotation += np.array([rx, ry, rz], dtype=np.float32)
        self.set_dirty()

    def scale_by(self, sx, sy, sz):
        self.scale *= np.array([sx, sy, sz], dtype=np.float32)
        self.set_dirty()

    def set_position(self, x, y, z):
        self.position = np.array([x, y, z], dtype=np.float32)
        self.set_dirty()

    def set_rotation(self, rx, ry, rz):
        self.rotation = np.array([rx, ry, rz], dtype=np.float32)
        self.set_dirty()

    def set_scale(self, sx, sy, sz):
        self.scale = np.array([sx, sy, sz], dtype=np.float32)
        self.set_dirty()

    def look_at(self, target):
        direction = np.asarray(target, dtype=np.float32) - self.position
        direction = direction / (np.linalg.norm(direction) + 1e-8)
        yaw = np.degrees(np.arctan2(direction[2], direction[0]))
        pitch = np.degrees(np.arcsin(direction[1]))
        self.rotation = np.array([pitch, yaw, 0.0], dtype=np.float32)
        self.set_dirty()

    def get_world_position(self):
        return self._world_matrix[:3, 3]

    def get_bounding_box(self):
        if self.mesh is None:
            bb_min = self.position - np.abs(self.scale)
            bb_max = self.position + np.abs(self.scale)
            return bb_min, bb_max
        local_min, local_max = self.mesh.get_bounding_box()
        corners = np.array([
            [local_min[0], local_min[1], local_min[2]],
            [local_max[0], local_min[1], local_min[2]],
            [local_min[0], local_max[1], local_min[2]],
            [local_max[0], local_max[1], local_min[2]],
            [local_min[0], local_min[1], local_max[2]],
            [local_max[0], local_min[1], local_max[2]],
            [local_min[0], local_max[1], local_max[2]],
            [local_max[0], local_max[1], local_max[2]],
        ])
        world_corners = []
        for c in corners:
            p = np.append(c, 1.0)
            wp = self._world_matrix @ p
            world_corners.append(wp[:3])
        world_corners = np.array(world_corners)
        return np.min(world_corners, axis=0), np.max(world_corners, axis=0)

    def clone(self, recursive=True):
        node = SceneNode(self.name + "_copy")
        node.position = self.position.copy()
        node.rotation = self.rotation.copy()
        node.scale = self.scale.copy()
        node.visible = self.visible
        node.locked = self.locked
        if self.mesh is not None:
            node.mesh = self.mesh.clone()
        if self.material is not None:
            node.material = self.material.clone()
        if recursive:
            for child in self.children:
                node.add_child(child.clone(True))
        return node


class Scene:
    def __init__(self, name="Scene"):
        self.name = name
        self.root = SceneNode("Root")
        self.cameras = []
        self.lights = []
        self.selected_node = None
        self.background_color = np.array([0.15, 0.15, 0.2, 1.0], dtype=np.float32)
        self.grid_enabled = True
        self.grid_size = 20.0
        self.grid_subdivisions = 20

    def add_node(self, node, parent=None):
        if parent is None:
            parent = self.root
        parent.add_child(node)
        return node

    def remove_node(self, node):
        if node.parent is not None:
            node.parent.remove_child(node)

    def get_all_nodes(self):
        nodes = []
        self._collect_nodes(self.root, nodes)
        return nodes

    def _collect_nodes(self, node, result):
        for child in node.children:
            result.append(child)
            self._collect_nodes(child, result)

    def find_node(self, name):
        for node in self.get_all_nodes():
            if node.name == name:
                return node
        return None

    def update(self):
        self.root.update_matrices()

    def add_camera(self, camera=None):
        if camera is None:
            from .camera import Camera
            camera = Camera("Camera_" + str(len(self.cameras)))
        self.cameras.append(camera)
        return camera

    def add_light(self, light=None):
        if light is None:
            from .light import Light
            light = Light("Light_" + str(len(self.lights)))
        self.lights.append(light)
        return light

    def select(self, node):
        self.selected_node = node

    def deselect(self):
        self.selected_node = None

    def duplicate_selected(self):
        if self.selected_node is not None:
            dup = self.selected_node.clone()
            dup.position[0] += 1.0
            self.add_node(dup)
            self.select(dup)
            return dup
        return None

    def delete_selected(self):
        if self.selected_node is not None:
            parent = self.selected_node.parent or self.root
            self.remove_node(self.selected_node)
            self.selected_node = None

    def clear(self):
        self.root = SceneNode("Root")
        self.cameras.clear()
        self.lights.clear()
        self.selected_node = None

    def get_node_count(self):
        return len(self.get_all_nodes())

    def raycast(self, origin, direction):
        origin = np.asarray(origin, dtype=np.float32)
        direction = np.asarray(direction, dtype=np.float32)
        closest_hit = None
        closest_dist = float('inf')
        for node in self.get_all_nodes():
            if node.mesh is None or not node.visible:
                continue
            hit_dist = self._raycast_node(origin, direction, node)
            if hit_dist is not None and hit_dist < closest_dist:
                closest_dist = hit_dist
                closest_hit = node
        if closest_hit is not None:
            return closest_hit, closest_dist
        return None, None

    def _raycast_node(self, origin, direction, node):
        bb_min, bb_max = node.get_bounding_box()
        inv_dir = 1.0 / (direction + 1e-8)
        t1 = (bb_min - origin) * inv_dir
        t2 = (bb_max - origin) * inv_dir
        tmin = np.minimum(t1, t2)
        tmax = np.maximum(t1, t2)
        t_near = np.max(tmin)
        t_far = np.min(tmax)
        if t_near <= t_far and t_far > 0:
            return t_near if t_near > 0 else t_far
        return None
